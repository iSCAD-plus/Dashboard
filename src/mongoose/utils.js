export const required = type => ({
  type,
  required: true,
});

export const optionalEnum = values => ({
  type: String,
  enum: {
    values,
    message: 'enum validator failed for path `{PATH}` with value `{VALUE}`.',
  },
});

export const requiredEnum = values => ({
  type: String,
  required: true,
  enum: {
    values,
    message: 'enum validator failed for path `{PATH}` with value `{VALUE}`.',
  },
});
