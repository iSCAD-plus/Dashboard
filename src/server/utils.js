import R from 'ramda';
import S from 'sanctuary';
import { simpleClique } from 'cliquer';

const prepend = R.ifElse(
  R.test(/^measure/),
  R.concat('$measures.'),
  R.concat('$')
);

const aggregationOperator = R.cond([
  [R.equals('count'), R.always({ $sum: 1 })],
  [R.equals('month'), R.always({ $first: { $month: prepend('date') } })],
  [R.T, key => ({ $first: prepend(key) })],
]);

const formId = R.cond([
  [R.equals('count'), R.always('')],
  [
    R.equals('month'),
    R.always({ $substr: [{ $month: prepend('date') }, 0, -1] }),
  ],
  [R.T, key => ({ $substr: [prepend(key), 0, -1] })],
]);

const generateId = S.pipe([
  R.map(formId),
  R.intersperse('|'),
  R.objOf('$concat'),
]);

export const getQueryKeys = S.pipe([
  R.path(['operation', 'selectionSet', 'selections']),
  R.head,
  R.path(['selectionSet', 'selections']),
  R.map(R.path(['name', 'value'])),
]);

export const createGroupAggregation = keys => ({
  $group: {
    ...simpleClique(aggregationOperator, keys),
    _id: generateId(keys),
  },
});

export const hasMeasurePrefix = R.any(k => k.startsWith('measure'));
