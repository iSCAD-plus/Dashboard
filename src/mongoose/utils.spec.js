import { required, optionalEnum, requiredEnum } from './utils';

it('#require', () => {
  expect(required('toast')).toMatchSnapshot();
});

const values = [1, 2, 3];
it('#optionalEnum', () => {
  expect(optionalEnum(values)).toMatchSnapshot();
});

it('#requiredEnum', () => {
  expect(requiredEnum(values)).toMatchSnapshot();
});
