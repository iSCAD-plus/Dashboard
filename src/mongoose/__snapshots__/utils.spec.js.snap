// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#optionalEnum 1`] = `
Object {
  "enum": Object {
    "message": "enum validator failed for path \`{PATH}\` with value \`{VALUE}\`.",
    "values": Array [
      1,
      2,
      3,
    ],
  },
  "type": [Function],
}
`;

exports[`#require 1`] = `
Object {
  "required": true,
  "type": "toast",
}
`;

exports[`#requiredEnum 1`] = `
Object {
  "enum": Object {
    "message": "enum validator failed for path \`{PATH}\` with value \`{VALUE}\`.",
    "values": Array [
      1,
      2,
      3,
    ],
  },
  "required": true,
  "type": [Function],
}
`;
